Db data set

JorisGoosen opened the pull request:
Implements https://github.com/jasp-stats/INTERNAL-jasp/issues/1194

JorisGoosen committed:
Sqlite as data backend.

Squashed lots of commits:

starting direct sqlite3 api usage

database is created with tables

create dataset now inserts a dummy row in the table

inserting works

loading dataset info also works

dataset syncs values with DB now

moved emptyvalues stuff from datasetpackage to separate class, now a child of DataSet
database-json and datafilepath are also now stored there.

Give Filter its own class and sync with DB

filtervalues are now also in DB

clean split between SQL code and c++ api

reorder name components

how to store data in the dataset?

make sure we use transactions

also more functionality and storage for columns and filters.
They are now both part of the DataSet_# tables which means they all have the same amount of rows by definition

more filter functionality and settled on "constructorJson/R"

move filter to CommonData

moved Filter to DataSetPackage to avoid destruction with DataSet

columns are indexed so there should be support for order...

better make some extra classses and rename the old ones

rename old dataset, column(s) to *SM

rename Label(s) to Label(s)SM

add empty new DataSet and Column class

values and filter are now synched to db

better OOP adherence

add some functions from ColumnSM but they wont compile ;)

more labelstuff

ok time to stop

reimagine what is supposed to be stored in a label class

and partially rewrote that

compiles but crashes

now it also works ^^

labels order is now reflected to database as well as label.label for ints

rewrite engine to use database, doesnt work at all yet

engine can load the data!

remove shared mem classes

try to get everything to compile after removing those classes

compiles and crashes

I guess I should commit this broken stuff before breaking it some more

actually set values...

loading data works ^^

add revisions to DataSet, Columns and Filters tables.

This allows for granular reloadingh (granular in that not *all* columns might need to be reloaded)

data loading and using goes pretty well now

Filter is no longer "singletonized" but created and destroyed with the DataSet
loading a second dataset after closing a first crashes pretty bad though...

Loading, closing and reloading now works!

Also make sure to not delete the database while jasp is still using it...

wip

some more code

it works now and sped up greatyl

add bunch a timers to ColumnUtils and some optimization tweaks that probably dont help very much

at least it works now and engine stops sleeping

filterWrite could be quicker (this way I hope)

write all values at once, should be much faster

put a max around value going into viewportH and V cause the height can become insanely large when closing/opening dataset.
This causes it to create items for all entries in the data. Which is quite a lot when you have 10cols and 1 million rows for instance...

batched loading also added

does however crash pretty bad on release mode

even iets anders

iterating

loading labels doesnt crash, editing works smoother but nominal data gets read wrong

tsja... wat is er mis met de nominals?

debug stuff for nominal error ?

remove debulogs and dont +1 rownumber as it is 0-based anyway

filter now works as it did with shared memory

Dont send filter result through shared memory but wreite directly to database

remove cruft
fix https://github.com/jasp-stats/jasp-issues/issues/1576 (value column too thin in variableswindow)
editing now seems to work smoothly

try to convert back to scalar if changed at a reasonable moment

	show filtered rowNumber and shutdown JASP cleanly

pointer rewrite

thinking hard on subNodeModels but ...

lots of changes, internal structure used for ModelIndex. Does not compile

it compiles and sort of works, but sort of in the sense it shows some data. not nearly everything though?

Fixed it, data shows

Editing now mostly works.

resizing data does not

Resizing data works ok now, computed columns still mess it up though

improve handling of variableswindow on nominal->scale conversion while open
also remove ComputedColumns refs to start writing that out of JASP
does not compile

try move all computedColumn stuff to Column etc

moved _revision to DataSetBaseNode for PackageModified feedback
moving more computedcolumns stuff to Column/DataSetPackage

almost compiles ^^

some more

it compiles again

computed columns almost work.
In that they firest work but then overwrite themselves with nonsense for some reason...

unselect computed column if it gets removed

changing columntype doesnt change columntype of dependents anymore

computedcolumns seem to function again

possibly improved things but still something goes wrong recovering a computed column from a previous error

computed columns keep working after an error now

removing columns through resizing now noticed by computed columns

make sure computed columns also function in datamode

something else

bit better

dont need _edit

I really should read the whole api documentation

steps forwards and backwards

improved some dialogs and RectangularButton and derivatives now show hovered color when activeFocussed (for keyboard nav)

use jaspTheme.fontCode in computedColumn R Window

dialog fixes

constructorjson isnt being reset anymoreo

Fix resizing dataset and remove bunch of debuglogging

make sure to also set _stopProcessing in stopEngines and co

rowNumber/primary keys in sqlite ARE 1-based after all
Also make sure we can resize the data to being smaller than it already was :p

make sure computed column creation type selection work and columnheaders and rownumbers in datasetview really should neatly line up with them lines

kapot

Handling removed columns better in constructor computed cols

preparing rewrite of saving and loading jaspfiles

preparing prettier exportercode

exporting new jasp archive format really cleans up the code in JASPExporter

remove weird java style manifest and replace by json for possible later expansion

does not compile

loads new jasp file and shows data

also load filter please

also load computed columns

klaar voor vandaag

show progress

make sure computed columns load (from penguins.jasp in any case)

remove some logging and small fixes

start documenting shit

more commentary

show db structure

make it compile on linux

Fixes issues with missing values in scalar columns

Added some workaround functions to allow storing NAN, INF and -INF as strings in double columns
This works because sqlite doesnt care (usually) what you put in each column
And it is necessary because it does NOT support NAN and (-)INF and just converts it to nothing...

put emptyvalues in emptyvalues

and databaseJson in databaseJson instead of mixing them up...

Suppress _info redefinition in columnsModel.cpp

JorisGoosen commented:
So reviews would be nice ^^

JorisGoosen commented:
@RensDofferhoff if I load debugdataset and check whether the missing values are correct:
<img width="327" alt="image" src="https://github.com/jasp-stats/jasp-desktop/assets/29062713/60544790-ddd0-42b7-8b6c-52e25f8f6457">

So im not sure why it isnt working with your compile?
Did you manually make the data as opposed to loading it maybe?

JorisGoosen commented:
Ah yes 
<img width="689" alt="image" src="https://github.com/jasp-stats/jasp-desktop/assets/29062713/8c255f73-4a08-4e48-9d1a-70becffa3d34">

JorisGoosen commented:
Fixed it
<img width="867" alt="afbeelding" src="https://github.com/jasp-stats/jasp-desktop/assets/29062713/06bf201e-7c13-4bd0-909e-84fbe1063d9d">

Problem was that SQLITE couldnt properly store nan/inf

boutinb commented on Tools/CMake/Modules.cmake:
Set right alphabetical order

boutinb commented on QMLComponents/controls/jaspcontrol.h:
```diff
@@ -10,7 +10,7 @@
 class JASPListControl;
 class BoundControl;
-class ComputedColumn;
+class Column;
```
Forward def not needed anymore apparently

boutinb commented on QMLComponents/components/JASP/Theme/Theme.qml:
```diff
@@ -61,7 +61,7 @@ JaspTheme
 
 	fontCode.family:									preferencesModel.codeFont
-	fontCode.pixelSize:									Math.round(10 * uiScale)
+	fontCode.pixelSize:									Math.round(11 * uiScale)
```
Why?

JorisGoosen replied:
Because it was supertiny

JorisGoosen commented on Tools/CMake/Modules.cmake:
```diff
@@ -19,38 +19,39 @@
+  "jaspMachineLearning"
+  "jaspMetaAnalysis"
+  "jaspNetwork"
+  "jaspQualityControl"
+  "jaspPower"
+  "jaspPredictiveAnalytics"
+  "jaspProphet"
+  "jaspReliability"
```
```suggestion
  "jaspPower"
  "jaspPredictiveAnalytics"
  "jaspProphet"
  "jaspQualityControl"
  "jaspReliability"
```

boutinb commented:
Add row on the last row, or add column after the last column or before the first column crash

boutinb commented:
I can open a file of 250 MB without problem (it is quicker than in 0.17.1), but closing the dataset is really long. I think the dbDelete function can be more efficient.

JorisGoosen commented:
> I can open a file of 250 MB without problem (it is quicker than in 0.17.1), but closing the dataset is really long. I think the dbDelete function can be more efficient.

Good point, I didnt really look into this (just killed JASP every time so didnt really notice :p)

boutinb commented:
This branch could not be compiled anymore. I've just fixed it ([Suppress _info redefinition in columnsModel.cpp](https://github.com/jasp-stats/jasp-desktop/pull/5020/commits/16da84b813a208e82760423f677374a1622f7512), but now JASP crashes when it starts...

JorisGoosen commented:
Ah you should not have removed the definition in the cpp.
Instead you should have made it static again in ColumnsModel.h

boutinb commented:
> Ah you should not have removed the definition in the cpp.
> Instead you should have made it static again in ColumnsModel.h

I set it as it is in stable.

JorisGoosen committed:
fix crash

JorisGoosen committed:
inserting columns and rows should now work regardless of where

JorisGoosen committed:
more timers and a speedup

JorisGoosen committed:
more timers

JorisGoosen commented:
Ah exporting to csv also seems to have a small problem.
And the crash on insert is only for loaded datasets

JorisGoosen committed:
Ok, the UNIQUE constraint messes up updating column-indices (as the order isnt fixed in sql update)
also, the new column to insert ought to get the size of the dataset as empty rows

Keep in mind that inserting rows in a nominal text column can destroy the entire column for some reason...

boutinb commented:
I can now add columns on an existing datafile.
But adding a row still deletes values of some columns

JorisGoosen committed:
make sure newly inserted columns are scalar and empty by default
also makes sure inserting rows works and doesnt erase all column(type)s

JorisGoosen commented:
New commit should fix that ^^

Ill have a look at dbDelete

JorisGoosen committed:
remove superfluous forward declaration

JorisGoosen committed:
progress for each row can be appreciated when loading large datasets

JorisGoosen commented:
Reminder to myself: be sure to check loading from an external database

JorisGoosen committed:
engines are loading data slightly less opaquely

it would be nice to have a progressbar in the results
And an update/change might be needed for the engine startup/kill protocol

boutinb commented:
Work better, but deleting a row remove data in some columns.

JorisGoosen commented:
Haha, well then I know what I will start with tomorrow :p

JorisGoosen committed:
use static stringvec to make const ref possible for allValuesAsStrings in InitColumn
Also fix a small problem in labeloriginalvalue retrieval
but rowInsert is still broken

JorisGoosen committed:
makes deleting a row work without messing up shit

JorisGoosen committed:
delete row like insert row

JorisGoosen committed:
make sure engine doesnt get into a reload->resume->reload loop for no good reason at all if the dataset is so small the desktop doesnt even register the reloadData response

JorisGoosen committed:
better default title for relaxing the impatient user

JorisGoosen commented:
Hmm, you cant actually click a cell twice for editing, you first need to click something else...

JorisGoosen committed:
memory optimizations

JorisGoosen committed:
cleaner csvimporter

JorisGoosen committed:
more fruitless attempts

RensDofferhoff committed:
safe some memory by determining the amount of rows in csv

RensDofferhoff committed:
use swap trick to dealloc vector to safe mem

JorisGoosen committed:
better editing

JorisGoosen committed:
even smoother

boutinb committed:
Add icons to popup menu

JorisGoosen committed:
add icons to header cutcopypaste

JorisGoosen committed:
greatly speed up single cell edits
more timers and dynamic_casts are better than static because some pointers might have become crap apparently

JorisGoosen committed:
more data buttons, bit crashy though

JorisGoosen committed:
buttons do something, but one of those things is crashing jasp

JorisGoosen committed:
buttons now work without crashing jasp

JorisGoosen committed:
support qt insert remove api better and stop crash on exit

JorisGoosen committed:
wip use proper qabstractitemmodel signals

reset dataset if all rows/columns are deleted

add options for creating computed columns, rudimentary and still crashes on some actions

adding computed columns opens them and deleting also works correctly now

use qt api more closely

but inserting computed columns still has a bit of a problem

try to stop it crashing...

at least it stopped crashing, but the order of columns is now messed up

sort of right order

ok no more crash I hope?
It is time to stop in any case

handle not removing all rows columns again

there are still some issues but im stopping for today

bunch of fixes

JorisGoosen committed:
make sure to clean up db at close workspace
also setModified in setData

JorisGoosen committed:
make edit item better behaved ^^

JorisGoosen committed:
remove finishEdit

JorisGoosen committed:
fixes for arrow navigation, itemEditable and column/row deletion is now prevented after selection is removed

JorisGoosen committed:
fix row delete selection behaviour

boutinb commented on Desktop/qquick/datasetview.cpp:
```diff
@@ -884,39 +903,59 @@ void DataSetView::updateExtraColumnItem()
+		size_t col=_prevEditCol, row=_prevEditRow;
+
+		//A delay might help the focus problem? No it doesnt...
+		QTimer::singleShot(10, _cellTextItems[col][row]->item, [col, row, this](){ _cellTextItems[col][row]->item->forceActiveFocus(); });
```
```suggestion
		QTimer::singleShot(10, _cellTextItems[col][row]->item, [col, row, this](){ if (_cellTextItems.contains(col) && _cellTextItems[col].contains(row) && _cellTextItems[col][row]->item) _cellTextItems[col][row]->item->forceActiveFocus(); });
```

boutinb committed:
Update Desktop/qquick/datasetview.cpp

boutinb commented:
Let's merge it!
